python/mspasspy/db/database.py

```diff
@@ -4752,3 +5026,333 @@ def save_textfile(
+        network="",
+        station="",
+        channel="",
+        location="",
+        duration=-1,
+        t0shift=0,
+    ):
+        """
+        A helper function to download the miniseed file from AWS s3.
+        A lambda function will be called, and do the timewindowing on cloud. The output file
+        of timewindow will then be downloaded and parsed by obspy.
+        Finally return an obspy stream object.
+
+        :param aws_access_key_id & aws_secret_access_key: credential for aws, used to initialize lambda_client
+        :param year:  year for the query mseed file(4 digit).
+        :param day_of_year:  day of year for the query of mseed file(3 digit [001-366])
+        :param network:  network code
+        :param station:  station code
+        :param channel:  channel code
+        :param location:  location code
+        :param duration:  window duration, default value is -1, which means no window will be performed
+        :param t0shift: shift the start time, default is 0
+        """
+        lambda_client = boto3.client(
+            service_name="lambda",
+            region_name="us-west-2",
+            aws_access_key_id=aws_access_key_id,
+            aws_secret_access_key=aws_secret_access_key,
+        )
+        s3_input_bucket = "scedc-pds"
+        s3_output_bucket = "scedcdata"  #   The output file can be saved to this bucket, currently not in use
+        year = str(year)
+        day_of_year = str(day_of_year)
+        if len(day_of_year) < 3:
+            day_of_year = "0" * (3 - len(day_of_year)) + day_of_year
+        source_key = (
+            "continuous_waveforms/" + year + "/" + year + "_" + day_of_year + "/"
+        )
+        if len(network) < 2:
+            network += "_" * (2 - len(network))
+        if len(station) < 5:
+            station += "_" * (5 - len(station))
+        if len(channel) < 3:
+            channel += "_" * (3 - len(channel))
+        if len(location) < 2:
+            location += "_" * (2 - len(location))
+
+        mseed_file = (
+            network + station + channel + location + "_" + year + day_of_year + ".ms"
+        )
+        source_key += mseed_file
+
+        event = {
+            "s3_input_bucket": s3_input_bucket,
+            "s3_output_bucket": s3_output_bucket,
+            "source_key": source_key,
+            "duration": duration,
+            "t0shift": t0shift,
+            "save_to_s3": False,
+        }
+
+        response = lambda_client.invoke(
+            FunctionName="TimeWindowFunction",  #   The name of lambda function on cloud
+            InvocationType="RequestResponse",
+            LogType="Tail",
+            Payload=json.dumps(event),
+        )
+
+        response_payload = json.loads(response["Payload"].read())
+        ret_type = response_payload["ret_type"]
+
+        if (
+            ret_type == "key"
+        ):  #   The output file is saved to another bucket (s3_output_bucket). We don't support this workflow now.
```

Comment:
I thought we have tested with large return object before, and why is this not supported here? Is there a newer version of the implementation somewhere?